src/batchmap.rs

```diff
@@ -0,0 +1,627 @@
+        }
+
+        assert_eq!(responses.len(), 2, "Expected two message from server");
+        assert_eq!(&responses[0].id, "1");
+        assert_eq!(&responses[1].id, "2");
+
+        shutdown_tx
+            .send(())
+            .expect("Sending shutdown signal to gRPC server");
+        tokio::time::sleep(Duration::from_millis(50)).await;
+        assert!(task.is_finished(), "gRPC server is still running");
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn error_length() -> Result<(), Box<dyn Error>> {
+        struct Logger;
+        #[tonic::async_trait]
+        impl batchmap::BatchMapper for Logger {
+            async fn batchmap(&self, mut input: Receiver<Datum>) -> Vec<BatchResponse> {
+                let responses: Vec<BatchResponse> = Vec::new();
+                while let Some(_datum) = input.recv().await {}
+                responses
+            }
+        }
+
+        let tmp_dir = TempDir::new()?;
+        let sock_file = tmp_dir.path().join("batchmap.sock");
+        let server_info_file = tmp_dir.path().join("batchmapper-server-info");
+
+        let mut server = batchmap::Server::new(Logger)
+            .with_server_info_file(&server_info_file)
+            .with_socket_file(&sock_file)
+            .with_max_message_size(10240);
+
+        assert_eq!(server.max_message_size(), 10240);
+        assert_eq!(server.server_info_file(), server_info_file);
+        assert_eq!(server.socket_file(), sock_file);
+
+        let (_shutdown_tx, shutdown_rx) = oneshot::channel();
+        let task = tokio::spawn(async move { server.start_with_shutdown(shutdown_rx).await });
+
+        tokio::time::sleep(Duration::from_millis(50)).await;
+
+        // https://github.com/hyperium/tonic/blob/master/examples/src/uds/client.rs
+        let channel = tonic::transport::Endpoint::try_from("http://[::]:50051")?
+            .connect_with_connector(service_fn(move |_: Uri| {
+                // https://rust-lang.github.io/async-book/03_async_await/01_chapter.html#async-lifetimes
+                let sock_file = sock_file.clone();
+                async move {
+                    Ok::<_, std::io::Error>(hyper_util::rt::TokioIo::new(
+                        UnixStream::connect(sock_file).await?,
+                    ))
+                }
+            }))
+            .await?;
+
+        let mut client = BatchMapClient::new(channel);
+        let request = batchmap::proto::BatchMapRequest {
+            keys: vec!["first".into(), "second".into()],
+            value: "hello".into(),
+            watermark: Some(prost_types::Timestamp::default()),
+            event_time: Some(prost_types::Timestamp::default()),
+            id: "1".to_string(),
+            headers: Default::default(),
+        };
+
+        let resp = client
+            .batch_map_fn(tokio_stream::iter(vec![request]))
+            .await?;
+        let mut r = resp.into_inner();
+
+        let mut error_flag = false;
+
+        if let Err(e) = r.message().await {
```

Comment:
You can return an `Err` to fail the test.
With an early return, you can avoid the need for an `error_flag` flag.
```rust
  let Err(server_err) = r.message().await else {
      return Err("Expected error from server".into());
  };

  assert_eq!(server_err.code(), tonic::Code::Internal);
  assert!(server_err.message().contains(
      "number of responses does not \
              match the number of messages received"
  ));
```